### PR TITLE
pkg/ubasic: add support for BASIC's "shell"  statement. 

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

A joke can go wrong when people don't understand it, and it can _wronger_ if they take it seriously...

This commit is based on #11319 and adds support for the "shell" statement. In BASIC, this is supposed to call a system command:

```BASIC
10 shell "ls $HOME"
```

I used "system()" to implement it. For more information regarding `shell` in other BASIC dialects, see:

- FreeBASIC: https://freebasic.net/wiki/wikka.php?wakka=KeyPgShell
- Gambas: http://gambaswiki.org/wiki/lang/shell
- Other: https://rosettacode.org/wiki/Execute_a_system_command#BASIC

### Testing procedure

Run the test. Test case no.3 takes ages in a board (I tested samr21) so you can use the following patch to disable it (place it under `pkg/ubasic/patches`).
`0006-remove-me-relax-test-2.patch`
```patch
From 83837232900bbdd2e8827046a428ecdb0655eb3c Mon Sep 17 00:00:00 2001
From: Juan Carrano <j.carrano@fu-berlin.de>
Date: Fri, 5 Apr 2019 16:29:48 +0200
Subject: [PATCH 6/6] remove me - relax test 2

---
 tests.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/tests.c b/tests.c
index b4ac79a..7c0575a 100644
--- a/tests.c
+++ b/tests.c
@@ -48,8 +48,8 @@ static const char program_goto[] =
 70 end\n";
 
 static const char program_loop[] =
-"10 for i = 0 to 126\n\
-20 for j = 0 to 126\n\
+"10 for i = 0 to 1\n\
+20 for j = 0 to 1\n\
 30 for k = 0 to 10\n\
 40 let a = i * j * k\n\
 rem 45 print a, i, j, k\n\
@@ -128,7 +128,7 @@ main(void)
   assert(ubasic_get_variable(2) == 108);
 
   run(program_loop);
-  assert(ubasic_get_variable(0) == (VARIABLE_TYPE)(126 * 126 * 10));
+   /* assert(ubasic_get_variable(0) == (VARIABLE_TYPE)(126 * 126 * 10)); */
 
   run(program_shell);
 
-- 
2.21.0
```

### Side notes.

The modifications to ubasic were unexpectedly easy. I spent most of the time figuring out how to get system() running in RIOT. One of the biggest issues I had was that at some point I added a main.c to the test application and afterwards remove it, bit the main.o was still there messing up the symbols!!!

### Issues/PRs references

Depends (and extends) #11319 .

### CODE REVIEW

**Comment by miri64** [comment]
> This commit is based on #11319 and adds support for the "shell" statement. In BASIC, this is supposed to call a system command.

Can you give a ref for that? "basic shell statement" is a bit frustrating to google...

**Comment by jcarrano** [comment]
> Can you give a ref for that? 

Well, to be honest, I'm not very sure of what I did. There are many BASIC dialects (uBASIC being one of them), but several seem to have a shell command that takes a single string as an argument and executes it in a sub-process using the system shell.

**Comment by jcarrano** [comment]
I updated the description with reference to other dialects' `shell`.

**Comment by miri64** [comment]
> I updated the description with reference to other dialects' shell.
>

Thank you!

File: sys/shell/commands/sc_sys.c [hunk 1]
```diff
@@ -29,3 +30,18 @@ int _reboot_handler(int argc, char **argv)
 
     return 0;
 }
+
+int _echo_handler(int argc, char **argv)
```
Comment (miri64) [inline sys/shell/commands/sc_sys.c]: This belongs into its own PR.

File: sys/shell/shell.c [hunk 2]
```diff
@@ -304,3 +304,17 @@ void shell_run(const shell_command_t *shell_commands, char *line_buf, int len)
         print_prompt();
     }
 }
+
+#ifdef MODULE_SHELL_COMMANDS
+/* Provide standard system() function */
+int system(const char *command)
```
Comment (miri64) [inline sys/shell/shell.c]: This as well.

Also I think this rather belongs into the POSIX module and `shell_commands` shouldn't be a dependency.

Comment (jcarrano) [reply sys/shell/shell.c]: This should have its own PR. I would not put it under POSIX, since system() is C99/C89 and placing it is the easiest way to make sure system() is tied to shell_commands.

Comment (miri64) [reply sys/shell/shell.c]: > since system() is C99/C89 and placing it is the easiest way to make sure system() is tied to shell_commands.

You can have shell commands without having the module `shell_commands` (which provides a number of *default* shell commands), so I don't understand the requirement for `shell_commands`.

File: sys/shell/shell.c [hunk 3]
```diff
@@ -304,3 +304,17 @@ void shell_run(const shell_command_t *shell_commands, char *line_buf, int len)
...
+{
+    int len = strlen(command);
+    char tmpstring[len+1];
+
+    strncpy(tmpstring, command, len+1);
+
+    handle_input_line(_shell_command_list, tmpstring);
```
Comment (jcarrano) [inline sys/shell/shell.c]: I should have used NULL here, since the _shell_command_list is going to be searched anyways.
```suggestion
    handle_input_line(NULL, tmpstring);
```

File: sys/shell/shell.c (hunk 2 above)
Comment (jcarrano) [reply sys/shell/shell.c]: I you don't have shell_commands, then you don't have access to _any_ command from within system(), Because you have no way for your application to communicate a custom command list:

https://github.com/RIOT-OS/RIOT/blob/200c465f36ffee6b1958a59c80b8879d09ea1d2b/sys/shell/shell.c#L54-L62

Nothing will break, but no commands will be available.

In 2038, when we finally get XFAs, this won't be an issue (if we survive the UNIX timestamp overflow)

**Comment by miri64** [comment]
Needs rebase and split-up.

File: sys/shell/shell.c (hunk 2 above)
Comment (miri64) [reply sys/shell/shell.c]: > I you don't have shell_commands, then you don't have access to any command from within `system()`

Well, that seems to be more of an issue of the current implementation. So I see it as premature to establish a dependency that in reality should not be there (I would expect e.g. `system("udp …")` to work if I call it with `gnrc_networking`.

> In 2038, when we finally get XFAs, this won't be an issue (if we survive the UNIX timestamp overflow)

ztimer will handle the latter. ;-)

File: sys/shell/commands/sc_sys.c (hunk 1 above)
Comment (miri64) [reply sys/shell/commands/sc_sys.c]: Also, I noticed today, that this somewhat clashes with [a task in our tutorials](https://github.com/RIOT-OS/Tutorials/blob/master/task-02/README.md#task-21-a-simple-echo-command-handler)

**Comment by jcarrano** [comment]
@miri64 Maybe we can leave the echo commit out (i.e. not even PR it). The only reason I put it there was to be able to have at least one command to test (otherwise the only command available by default is "reboot").

The thing is, because of the way the shell works it not easy or practical to make the application specific commands available to `system()`. And because of how the shell is compiled, there is no [easy] way to add custom system commands without going into the shell module.

The result of all this is that it makes testing this feature harder.

**Comment by miri64** [comment]
Couldn't the `echo` command be included via a pseudo-module? This way it is a "system" command, but still optional.

**Comment by jcarrano** [comment]
@miri64 yes, that goes along the lines of what I was proposing in #9258, which is to allow choosing which system functionality (I will try to avoid calling them "system calls" because in RIOT they are not) are provided and to allow choosing an implementation.

The first question, then, is: do we want 'system()'. I think yes because it is in the standard and because it is useful.

**Comment by miri64** [comment]
> The first question, then, is: do we want 'system()'. I think yes because it is in the standard and because it is useful.

I'm not against it.

**Comment by jcarrano** [comment]
> I'm not against it.

Then I'll make a PR just for that.
